File: api/rest/node.go
package rest

import (
	"errors"
	"fmt"
	"github.com/musenwill/raftdemo/api"
	error2 "github.com/musenwill/raftdemo/api/error"
	"github.com/musenwill/raftdemo/fsm"
)

type NodeMgr struct {
	nodes map[string]fsm.Prober
}

func (p *NodeMgr) List() (api.ListResponse, *error2.HttpError) {
	result := api.ListResponse{}
	result.Total = len(p.nodes)

	for _, n := range p.nodes {
		result.Entries = append(result.Entries, api.Node{
			Host:          n.GetHost(),
			Term:          n.GetTerm(),
			State:         string(n.GetState()),
			CommitIndex:   n.GetCommitIndex(),
			LastAppliedID: n.GetLastAppliedIndex(),
			Leader:        n.GetLeader(),
			VoteFor:       n.GetVoteFor(),
			Logs:          n.GetLogs(),
		})
	}

	return result, nil
}

func (p *NodeMgr) Get(host string) (api.Node, *error2.HttpError) {
	node, ok := p.nodes[host]
	if !ok {
		return api.Node{}, error2.DataNotFoundError(errors.New(fmt.Sprintf("node %s not found", host)))
	}
	return api.Node{
		Host:          node.GetHost(),
		Term:          node.GetTerm(),
		State:         string(node.GetState()),
		CommitIndex:   node.GetCommitIndex(),
		LastAppliedID: node.GetLastAppliedIndex(),
		Leader:        "",
		VoteFor:       "",
		Logs:          node.GetLogs(),
	}, nil
}

File: frontend/mock/node.js
import Mock from 'mockjs'

const nodes = Mock.mock({
  'entries|3-5': [{
    'host|1': ['node-1', 'node-2', 'node-3', 'node-4'],
    'term|1': [1, 2, 3, 4],
    'state|1': ['Follower', 'Candidate', 'Leader'], 
    'commit_index|1': [-1, 0, 1, 2, 3],
    'last_applied_id|1': [-1, 0, 1, 2, 3],
    'leader|1': ['node-1', 'node2', 'node3', 'node4'],
    'vote_for|1': ['', '', '', '', '', '', '', 'node-1', 'node-2', 'node-3', 'node-4']
  }],
  'total|1-1': [2, 3, 4, 5],
})



export default [
  {
    url: '/v1/nodes',
    type: 'get',
    response: config => {
      return {
        total: nodes.total,
        entries: nodes.entries
      }
    }
  },
  {
    url: '/v1/nodes',
    type: 'put',
    response: config => {
      return {
      }
    }
  }
]
